### DISCUSSION TITLE
Hooks to execute arbitrary commands...?

### DISCUSSION THREAD

**Discussion by BoxOfSnoo** [garethgeorge/backrest#551] [Ideas]
I would like to sync a completed - and successful - backup to a secondary drive... one question was posted here: https://github.com/garethgeorge/backrest/discussions/438 seemingly without a final solution.

In my previous restic setup I was using `restic copy` to sync the repo to another. While not perfect, worked great and gave me an offsite backup that saved my bacon!

I suppose I could _explicitly_ write the full restic commands in a bash script, but I have a few gotchas...

1. The restic binary, while visible in the /bin image of the docker image, is versioned, so I'd have to first figure out the filename to execute that.  It would be nice if there was maybe a symlink - or even a template field that would expand to the restic version/filename.
2. The `Repo` struct is not fully documented... I discovered that `{{ Repo.Uri }}` nicely returns the path that I can include on the command line with the `-r` parameter.  However, the rest of the options are somewhat opaque to me.  Is there a way to get at the configured password, for example?

Does this sound reasonable? Or is there something I'm missing?  Thanks in advance.

**Comment by garethgeorge** [comment]
Hey, first off you're right that this is hard to do with backrest today.

The proto definition for the repo struct can be found here: https://github.com/garethgeorge/backrest/blob/main/proto/v1/config.proto#L34-L45 . I think it's not entirely unreasonable to go about using this to script restic.

The overall struct and utilities for the full set of variables available to hooks are defined here, https://github.com/garethgeorge/backrest/blob/main/internal/orchestrator/tasks/hookvars.go#L15-L29. This is a piece of code that's on my road map for some cleanup / much needed attention.

Re: the specific issue of running commands in hooks, I think it could be reasonable for backrest to provide a way of formatting a restic command w/options from the repo given the command line arguments. It's a bit out of scope for something I have time to look at implementing at the moment, but if anyone were motivated to implement it I'd be happy to discuss approaches in an issue. Minimally a helper that returns the path to the restic binary (returned by this utility https://github.com/garethgeorge/backrest/blob/b7abbf9603196e1ed169ebbe81d97d5fa89dea08/internal/resticinstaller/resticinstaller.go#L228) is something that could be added to the template function set.
